mamba_ssm/ops/selective_scan_interface_ compilable.py

```diff
@@ -0,0 +1,272 @@
+):
+    if u.stride(-1) != 1:
+        u = u.contiguous()
+    if delta.stride(-1) != 1:
+        delta = delta.contiguous()
+    if D is not None:
+        D = D.contiguous()
+    if B.stride(-1) != 1:
+        B = B.contiguous()
+    if C.stride(-1) != 1:
+        C = C.contiguous()
+    if z is not None and z.stride(-1) != 1:
+        z = z.contiguous()
+
+    squeeze_B = False
+    if B.dim() == 3:
+        B = rearrange(B, "b dstate l -> b 1 dstate l").contiguous()
+        squeeze_B = True
+
+    squeeze_C = False
+    if C.dim() == 3:
+        C = rearrange(C, "b dstate l -> b 1 dstate l").contiguous()
+        squeeze_C = True
+
+    out, x, *rest = selective_scan_cuda.fwd(u, delta, A, B, C, D, z, delta_bias, delta_softplus)
+    has_z = z is not None
+    final_out = rest[0].clone() if has_z else out.clone()
+    last_state = x[:, :, -1, 1::2].clone() if return_last_state else u.new_empty(0)
+    return final_out, last_state, out, x, squeeze_B, squeeze_C, has_z
+
+@torch.library.custom_op(
+    "custom_ops::selective_scan_bwd",
+    device_types=["cuda"],
+    mutates_args=(),
+)
+def custom_selective_scan_bwd(
+    dout: torch.Tensor,
+    u: torch.Tensor,
+    delta: torch.Tensor,
+    A: torch.Tensor,
+    B: torch.Tensor,
+    C: torch.Tensor,
+    D: Optional[torch.Tensor],
+    z: Optional[torch.Tensor],
+    delta_bias: Optional[torch.Tensor],
+    delta_softplus: bool,
+    out: torch.Tensor,
+    x: torch.Tensor,
+    squeeze_B: bool,
+    squeeze_C: bool,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    pass
+
+@torch.library.register_fake("custom_ops::selective_scan_bwd")
+def custom_selective_scan_bwd_fake(
+    dout,
+    u,
+    delta,
+    A,
+    B,
+    C,
+    D,
+    z,
+    delta_bias,
+    delta_softplus,
+    out,
+    x,
+    squeeze_B,
+    squeeze_C,
+):
+    du = torch.empty_like(u)
+    ddelta = torch.empty_like(delta)
+    dA = torch.empty_like(A)
+    dB = torch.empty_like(B)
+    dC = torch.empty_like(C)
+    dD = torch.empty_like(D) if (D is not None and D.numel() > 0) else u.new_empty(0)
+    dz = torch.empty_like(z) if (z is not None and z.numel() > 0) else u.new_empty(0)
+    ddelta_bias = torch.empty_like(delta_bias) if (delta_bias is not None and delta_bias.numel() > 0) else u.new_empty(0)
+    return du, ddelta, dA, dB, dC, dD, dz, ddelta_bias
+
+@torch.library.register_kernel("custom_ops::selective_scan_bwd", "cuda")
+def custom_selective_scan_bwd_cuda(
+    dout: torch.Tensor,
+    u: torch.Tensor,
+    delta: torch.Tensor,
+    A: torch.Tensor,
+    B: torch.Tensor,
+    C: torch.Tensor,
+    D: Optional[torch.Tensor],
+    z: Optional[torch.Tensor],
+    delta_bias: Optional[torch.Tensor],
+    delta_softplus: bool,
+    out: torch.Tensor,
+    x: torch.Tensor,
+    squeeze_B: bool,
+    squeeze_C: bool,
+):
+    if dout.stride(-1) != 1:
+        dout = dout.contiguous()
+    B = B.contiguous()
```

Comment:
We shouldn't have to do this again. Probably doesn't affect anything as torch has an internal flag for contiguous, but unneeded for both `B` and `C` as we save them in the backwards contexts.

Reply:
Yes probably it is not going to impact. But we could remove these